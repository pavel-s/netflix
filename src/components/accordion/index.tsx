import {
  createContext,
  Dispatch,
  SetStateAction,
  useContext,
  useState,
} from 'react';
import { TRSC } from '../../types';
import {
  Container,
  Inner,
  Body,
  Frame,
  Header,
  Item,
  Title,
} from './styles/accordion';

type TAccordion = TRSC<typeof Container> & {
  Title: TRSC<typeof Title>;
  Frame: TRSC<typeof Frame>;
  Item: TRSC<typeof Item>;
  Header: TRSC<typeof Header>;
  Body: TRSC<typeof Body>;
};

const Accordion: TAccordion = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

const ToggleContext = createContext<{
  show?: boolean;
  setShow?: Dispatch<SetStateAction<boolean>>;
}>({});

Accordion.Title = ({ children, ...rest }) => (
  <Title {...rest}>{children}</Title>
);

Accordion.Frame = ({ children, ...rest }) => (
  <Frame {...rest}>{children}</Frame>
);

Accordion.Item = function AccordionItem({ children, ...rest }) {
  const [show, setShow] = useState(false);

  return (
    <ToggleContext.Provider value={{ show, setShow }}>
      <Item {...rest}>{children}</Item>
    </ToggleContext.Provider>
  );
};

Accordion.Header = function AccordionHeader({ children, ...rest }) {
  const { show, setShow } = useContext(ToggleContext);

  const toggleShow = () => {
    setShow && setShow((show: boolean) => !show);
  };

  return (
    <Header onClick={toggleShow} {...rest}>
      {children}{' '}
      {show ? (
        <img src='/images/icons/close-slim.png' alt='' />
      ) : (
        <img src='/images/icons/add.png' alt='' />
      )}
    </Header>
  );
};

Accordion.Body = function AccordionBody({ children, ...rest }) {
  const { show } = useContext(ToggleContext);
  return (
    <Body className={show ? 'open' : 'closed'}>
      <span {...rest}>{children}</span>
    </Body>
  );
};

export default Accordion;
