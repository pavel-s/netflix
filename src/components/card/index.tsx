import { TRSC } from '../../types';

import {
  Container,
  CardContainer,
  Group,
  Title,
  Items,
  Item,
  ItemImage,
  Meta,
  Subtitle,
  Text,
  Feature,
  Maturity,
  FeatureClose,
} from './styles/card';

type TCard = TRSC<typeof Container> & {
  Container: TRSC<typeof Container>;
  Group: TRSC<typeof Group>;
  Title: TRSC<typeof Title>;
  Items: TRSC<typeof Items>;
  Item: TRSC<typeof Item>;
  ItemImage: TRSC<typeof ItemImage>;
  Meta: TRSC<typeof Meta>;
  Subtitle: TRSC<typeof Subtitle>;
  Text: TRSC<typeof Text>;
  Feature: TRSC<typeof Feature>;
  Maturity: TRSC<typeof Maturity>;
  FeatureClose: TRSC<typeof FeatureClose>;
};

const Card: TCard = ({ children, ...rest }) => {
  return <CardContainer {...rest}>{children}</CardContainer>;
};

Card.Container = function CardCardContainer({ children, ...rest }) {
  return <Container {...rest}>{children}</Container>;
};

Card.Group = function CardGroup({ children, ...rest }) {
  return <Group {...rest}>{children}</Group>;
};

Card.Title = function CardTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Card.Items = function CardItems({ children, ...rest }) {
  return <Items {...rest}>{children}</Items>;
};

Card.Item = function CardItem({ children, ...rest }) {
  return <Item {...rest}>{children}</Item>;
};

Card.ItemImage = function CardItemImage({ children, ...rest }) {
  return <ItemImage {...rest}>{children}</ItemImage>;
};

Card.Meta = function CardMeta({ children, ...rest }) {
  return <Meta {...rest}>{children}</Meta>;
};

Card.Subtitle = function CardSubtitle({ children, ...rest }) {
  return <Subtitle {...rest}>{children}</Subtitle>;
};

Card.Text = function CardText({ children, ...rest }) {
  return <Text {...rest}>{children}</Text>;
};

Card.Feature = function CardFeature({ children, ...rest }) {
  return <Feature {...rest}>{children}</Feature>;
};

Card.Maturity = function CardMaturity({ maturity, ...rest }) {
  return (
    <Maturity maturity={maturity} {...rest}>
      {maturity}
    </Maturity>
  );
};

Card.FeatureClose = function CardFeatureClose({ children, ...rest }) {
  return <FeatureClose {...rest}>{children}</FeatureClose>;
};

export default Card;
