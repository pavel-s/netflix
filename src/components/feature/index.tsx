import { TRSC } from '../../types';
import { Container, Inner, Title, Subtitle } from './styles/feature';

type TFeature = TRSC<'div'> & {
  Title: TRSC<'h1'>;
  Subtitle: TRSC<'h2'>;
};

const Feature: TFeature = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

Feature.Title = function FeatureTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Feature.Subtitle = function FeatureSubtitle({ children, ...rest }) {
  return <Subtitle {...rest}>{children}</Subtitle>;
};

export default Feature;
