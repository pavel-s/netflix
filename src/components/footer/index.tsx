import { TRSC } from '../../types';
import { Container, Inner, Title, Links, Link, Text } from './styles/footer';

type TFooter = TRSC<'div'> & {
  Title: TRSC<'p'>;
  Links: TRSC<'ul'>;
  Link: TRSC<'a'>;
  Text: TRSC<'p'>;
};

const Footer: TFooter = ({ children, ...rest }) => {
  return (
    <Container {...rest}>
      <Inner>{children}</Inner>
    </Container>
  );
};

Footer.Title = ({ children, ...rest }) => {
  return <Title {...rest}>{children}</Title>;
};

Footer.Links = ({ children, ...rest }) => {
  return <Links {...rest}>{children}</Links>;
};

Footer.Link = ({ children, ...rest }) => {
  return (
    <Link>
      <a {...rest}>{children}</a>
    </Link>
  );
};

Footer.Text = ({ children, ...rest }) => {
  return <Text {...rest}>{children}</Text>;
};

export default Footer;
