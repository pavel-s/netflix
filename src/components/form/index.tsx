import { TRSC } from '../../types';
import {
  Container,
  Base,
  Title,
  Input,
  Error,
  Submit,
  Row,
  Checkbox,
  Text,
  TextBig,
  Link,
  TextButton,
} from './styles/form';

type TForm = TRSC<typeof Container> & {
  Base: TRSC<typeof Base>;
  Title: TRSC<typeof Title>;
  Input: TRSC<typeof Input>;
  Error: TRSC<typeof Error>;
  Submit: TRSC<typeof Submit>;
  Row: TRSC<typeof Row>;
  Checkbox: TRSC<typeof Checkbox>;
  Text: TRSC<typeof Text>;
  TextBig: TRSC<typeof TextBig>;
  Link: TRSC<typeof Link>;
  TextButton: TRSC<typeof TextButton>;
};

const Form: TForm = ({ children, ...rest }) => {
  return <Container {...rest}>{children}</Container>;
};

Form.Base = function FormBase({ children, ...rest }) {
  return <Base {...rest}>{children}</Base>;
};

Form.Title = function FormTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Form.Input = function FormInput({ ...rest }) {
  return <Input {...rest} />;
};

Form.Error = function FormError({ ...rest }) {
  return <Error {...rest} />;
};

Form.Submit = function FormSubmit({ isLoading = false, children, ...rest }) {
  return (
    <Submit {...rest} isLoading={isLoading} type='submit'>
      {children}
    </Submit>
  );
};

Form.Row = function FormRow({ children, ...rest }) {
  return <Row {...rest}>{children}</Row>;
};

Form.Checkbox = function FormCheckbox({ children, ...rest }) {
  return <Checkbox {...rest}>{children}</Checkbox>;
};

Form.Text = function FormText({ children, ...rest }) {
  return <Text {...rest}>{children}</Text>;
};

Form.TextBig = function FormTextBig({ children, ...rest }) {
  return <TextBig {...rest}>{children}</TextBig>;
};

Form.Link = function FormLink({ children, ...rest }) {
  return <Link {...rest}>{children}</Link>;
};

Form.TextButton = function FormTextButton({ children, ...rest }) {
  return <TextButton {...rest}>{children}</TextButton>;
};

export default Form;
