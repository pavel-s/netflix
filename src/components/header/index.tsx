import { Background, Container, Logo, SignButton } from './styles/header';
import { LinkProps } from 'react-router-dom';
import { TRSC } from '../../types';

type THeader = TRSC<'div'> & {
  // Container: TRSC<'div'>;
  Logo: TRSC<'img'>;
  Frame: TRSC<'div'>;
  SignButton: TRSC<'a', LinkProps>;
};

const Header: THeader = ({ children, ...rest }) => {
  return <Background {...rest}>{children}</Background>;
};

Header.Frame = function HeaderFrame({ children, ...rest }) {
  return <Container {...rest}>{children}</Container>;
};

Header.Logo = function HeaderLogo({ children, ...rest }) {
  return <Logo {...rest} />;
};

Header.SignButton = function HeaderSignButton({ children, ...rest }) {
  return <SignButton {...rest}>{children}</SignButton>;
};

export default Header;
