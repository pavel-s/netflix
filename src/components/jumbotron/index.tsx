import { TRSC } from '../../types';
import {
  Container,
  Item,
  Inner,
  Pane,
  Title,
  Subtitle,
  Image,
  TFlexDirection,
} from './styles/jumbotron';

type TJumbotron = TRSC<typeof Item, { direction: TFlexDirection }> & {
  Container: TRSC<typeof Container>;
  Pane: TRSC<typeof Pane>;
  Title: TRSC<typeof Title>;
  Subtitle: TRSC<typeof Subtitle>;
  Image: TRSC<typeof Image>;
};

const Jumbotron: TJumbotron = ({ direction = 'row', children, ...rest }) => {
  return (
    <Item {...rest}>
      <Inner direction={direction}>{children}</Inner>
    </Item>
  );
};

Jumbotron.Container = function JumbotronContainer({ children, ...rest }) {
  return <Container {...rest}>{children}</Container>;
};

Jumbotron.Pane = function JumbotronPane({ direction, children, ...rest }) {
  return (
    <Pane direction={direction} {...rest}>
      {children}
    </Pane>
  );
};

Jumbotron.Title = function JumbotronTitle({ children, ...rest }) {
  return <Title {...rest}>{children}</Title>;
};

Jumbotron.Subtitle = function JumbotronSubtitle({ children, ...rest }) {
  return <Subtitle {...rest}>{children}</Subtitle>;
};

Jumbotron.Image = function JumbotronImage({ children, ...rest }) {
  return (
    <div>
      <Image {...rest} />
    </div>
  );
};

export default Jumbotron;
