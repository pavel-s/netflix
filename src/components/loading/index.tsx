import React from 'react';
import { TRSC } from '../../types';
import { Container, Image, LockBody, Spinner } from './styles/loading';

type TLoading = TRSC<
  typeof Image,
  Pick<typeof Spinner extends React.ComponentType<infer P> ? P : {}, 'size'>
>;

const Loading: TLoading = ({ size, src, ...rest }) => {
  return (
    <Container data-testid='loading'>
      <LockBody />
      <Spinner size={size}>
        {src ? <Image src={src} {...rest} /> : null}
      </Spinner>
    </Container>
  );
};

export default Loading;
