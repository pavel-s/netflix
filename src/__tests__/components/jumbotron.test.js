import { render } from '@testing-library/react';
import { Jumbotron } from '../../components';
import jumbotronData from './../../fixtures/jumbo.json';

describe('<Jumbotron />', () => {
  it('render Jumbotron with jumbotronData', () => {
    const { container, queryByText } = render(
      <Jumbotron.Container>
        {jumbotronData.map((item) => {
          const direction =
            item.direction === 'row-reverse' ? 'row-reverse' : 'row';
          return (
            <Jumbotron direction={direction} key={item.id}>
              <Jumbotron.Pane direction={direction}>
                <Jumbotron.Title>{item.title}</Jumbotron.Title>
                <Jumbotron.Subtitle>{item.subTitle}</Jumbotron.Subtitle>
              </Jumbotron.Pane>
              <Jumbotron.Pane direction={direction}>
                <Jumbotron.Image src={item.image} alt={item.alt} />
              </Jumbotron.Pane>
            </Jumbotron>
          );
        })}
      </Jumbotron.Container>
    );

    // title
    expect(queryByText(jumbotronData[0].title)).toBeTruthy();
    // subtitle
    expect(queryByText(jumbotronData[0].subTitle)).toBeTruthy();
    // image
    expect(document.querySelector('img')).toHaveAttribute(
      'src',
      jumbotronData[0].image
    );

    // all items rendered
    expect(container.firstChild.childNodes.length).toEqual(
      jumbotronData.length
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('render Jumbotron with jumbotronData with default direction', () => {
    const { container, queryByText } = render(
      <Jumbotron.Container>
        {jumbotronData.map((item) => {
          const direction =
            item.direction === 'row-reverse' ? 'row-reverse' : 'row';
          return (
            <Jumbotron key={item.id}>
              <Jumbotron.Pane>
                <Jumbotron.Title>{item.title}</Jumbotron.Title>
                <Jumbotron.Subtitle>{item.subTitle}</Jumbotron.Subtitle>
              </Jumbotron.Pane>
              <Jumbotron.Pane>
                <Jumbotron.Image src={item.image} alt={item.alt} />
              </Jumbotron.Pane>
            </Jumbotron>
          );
        })}
      </Jumbotron.Container>
    );

    // title
    expect(queryByText(jumbotronData[0].title)).toBeTruthy();
    // subtitle
    expect(queryByText(jumbotronData[0].subTitle)).toBeTruthy();
    // image
    expect(document.querySelector('img')).toHaveAttribute(
      'src',
      jumbotronData[0].image
    );

    // all items rendered
    expect(container.firstChild.childNodes.length).toEqual(
      jumbotronData.length
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
