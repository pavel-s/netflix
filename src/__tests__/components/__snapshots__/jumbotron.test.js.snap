// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Jumbotron /> render Jumbotron with jumbotronData 1`] = `
<div
  class="sc-bdnxRM"
>
  <div
    class="sc-gtsrHT czMKoh"
  >
    <div
      class="sc-dlnjwi hUlGbK"
      direction="row"
    >
      <div
        class="sc-hKFxyN gWLUzV"
        direction="row"
      >
        <h1
          class="sc-eCApnc efAkPC"
        >
          Enjoy on your TV.
        </h1>
        <h2
          class="sc-jSFjdj egzzfw"
        >
          Watch on smart TVs, PlayStation, Xbox, Chromecast, Apple TV, Blu-ray players and more.
        </h2>
      </div>
      <div
        class="sc-hKFxyN gWLUzV"
        direction="row"
      >
        <div>
          <img
            alt="Tiger King on Netflix"
            class="sc-gKAaRy jsKPnu"
            src="/images/misc/home-tv.jpg"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="sc-gtsrHT czMKoh"
  >
    <div
      class="sc-dlnjwi eVgYpP"
      direction="row-reverse"
    >
      <div
        class="sc-hKFxyN gWLUzV"
        direction="row-reverse"
      >
        <h1
          class="sc-eCApnc efAkPC"
        >
          Download your programmes to watch on the go.
        </h1>
        <h2
          class="sc-jSFjdj egzzfw"
        >
          Save your data and watch all your favourites offline.
        </h2>
      </div>
      <div
        class="sc-hKFxyN gWLUzV"
        direction="row-reverse"
      >
        <div>
          <img
            alt="Watch on mobile"
            class="sc-gKAaRy jsKPnu"
            src="/images/misc/home-mobile.jpg"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="sc-gtsrHT czMKoh"
  >
    <div
      class="sc-dlnjwi hUlGbK"
      direction="row"
    >
      <div
        class="sc-hKFxyN gWLUzV"
        direction="row"
      >
        <h1
          class="sc-eCApnc efAkPC"
        >
          Watch everywhere.
        </h1>
        <h2
          class="sc-jSFjdj egzzfw"
        >
          Stream unlimited films and TV programmes on your phone, tablet, laptop and TV without paying more.
        </h2>
      </div>
      <div
        class="sc-hKFxyN gWLUzV"
        direction="row"
      >
        <div>
          <img
            alt="Money Heist on Netflix"
            class="sc-gKAaRy jsKPnu"
            src="/images/misc/home-imac.jpg"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
